File: common/src/main/java/org/bread_experts_group/breadlib/task/TaskManager.java
package org.bread_experts_group.breadlib.task;

import org.bread_experts_group.breadlib.util.Util;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;
import java.util.function.Consumer;

public class TaskManager {
	public static final HashMap<Class<?>, ArrayList<Consumer<? extends Task>>> tasks = new HashMap<>();

	public static <T extends Task> void newTask(Class<T> tClass, Consumer<T> task) {
		var list = Util.getOrPut(tasks, tClass, ArrayList::new);
		list.add(task);
	}

	/**
	 * Runs all consumers registered to the specified task, then returns the task.
	 */
	@SuppressWarnings("unchecked")
	public static <T extends Task> T runTasks(T task) {
		ArrayList<Consumer<? extends Task>> list = null;
		for (Map.Entry<Class<?>, ArrayList<Consumer<? extends Task>>> entry : tasks.entrySet()) {
			if (entry.getKey().isAssignableFrom(task.getClass())) {
				list = entry.getValue();
				break;
			}
		}
		if (list == null || list.isEmpty()) return task;
		for (Consumer<? extends Task> consumer : list) ((Consumer<T>) consumer).accept(task);
		return task;
	}
}

File: common/src/main/kotlin/org/bread_experts_group/breadlib/task/TaskManager.kt
package org.bread_experts_group.breadlib.task

import java.util.function.Consumer

object TaskManager {
	val tasks: MutableMap<Class<*>, MutableList<Consumer<out Task>>> = mutableMapOf()

	fun <T : Task> newTask(tClass: Class<T>, task: Consumer<T>) {
		val list = tasks.getOrPut(tClass) { mutableListOf() }
		list.add(task)
	}

	/**
	 * Runs all consumers registered to the specified task, then returns the task.
	 */
	@JvmStatic
	@Suppress("UNCHECKED_CAST")
	fun <T : Task> runTasks(task: T): T {
		var list: MutableList<Consumer<out Task>> = mutableListOf()
		for ((key, value) in tasks) {
			if (key.isAssignableFrom(task.javaClass)) {
				list = value
				break
			}
		}
		if (list.isNotEmpty()) for (consumer in list) (consumer as Consumer<T>).accept(task)
		return task
	}

	inline fun <reified T : Task> newTask(noinline task: (T) -> Unit): Unit = newTask(T::class.java, task)
}